src/payment/views.py

```diff
@@ -0,0 +1,154 @@
+            price_id = (
+                product.monthly_price_id
+                if plan == "monthly"
+                else product.annual_price_id
+            )
+        except StripeProduct.DoesNotExist:
+            return Response({"error": "Product not found"}, status=404)
+
+        try:
+            checkout_session = stripe.checkout.Session.create(
+                payment_method_types=["card"],
+                customer_email=user.email,
+                line_items=[{"price": price_id, "quantity": 1}],
+                mode="subscription",
+                billing_address_collection="required",
+                currency="eur",
+                success_url=request.build_absolute_uri(
+                    reverse("v1:payment:subscription-success")
+                ),
+                cancel_url=request.build_absolute_uri(
+                    reverse("v1:payment:subscription-cancel")
+                ),
+                metadata={
+                    "user_id": user.id,
+                    "product_name": product.name,
+                },
+            )
+
+            data = {
+                **request.data,
+                "stripe_subscription_id": checkout_session.id,
+            }
+
+            headers = self.get_success_headers(data)
+            return Response(
+                {
+                    "sessionId": checkout_session.id,
+                    "checkout_url": checkout_session.url,
+                },
+                status=status.HTTP_201_CREATED,
+                headers=headers,
+            )
+        except Exception as e:
+            print("Error during session creation:", e)
+            return Response({"error": str(e)}, status=500)
+
+    def retrieve(self, request, *args, **kwargs):
+        obj = get_object_or_404(self.get_queryset(), user=request.user)
+        serializer = self.get_serializer(obj)
+        return Response(serializer.data)
+
+
+class SubscriptionUserCancelView(APIView):
+    """
+    Handles the cancellation of a subscription by the user.
+
+    This endpoint is used to cancel a subscription by the user.
+    It is called when the user requests to cancel their subscription.
+    """
+
+    def post(self, request, *args, **kwargs):
+        user = request.user
+        subscription = get_object_or_404(Subscription, user=user)
+        stripe.api_key = settings.STRIPE_API_KEY
+
+        try:
+            stripe.Subscription.modify(
+                subscription.stripe_subscription_id, cancel_at_period_end=True
+            )
+
+            subscription.cancel_at_period_end = True
+            subscription.save()
+
+            return Response(
+                {"message": "Subscription cancelled successfully"},
+                status=status.HTTP_200_OK,
+            )
+        except stripe.error.InvalidRequestError as e:
+            return Response(
+                {"error": f"Stripe error: {str(e)}"},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
```

Comment:
Let's narrow this down to the `stripe.Subscription.modify()` call only.
But I would be even tempted to keep it simple and not catch it. It would be logged into Sentry, frontend should have a generic error handling anyway.